add `_irred_abelian(G::Oscar.GAPGroup, F::FinField)`

ThomasBreuer opened the pull request:
This is analogous to the available `_irred_abelian(G::Oscar.GAPGroup)` which computes the `QQ`-irreducibles of finite abelian groups.

(The functionality will be mentioned in a forthcoming paper.)

ThomasBreuer committed:
add `_irred_abelian(G::Oscar.GAPGroup, F::FinField)`

This is analogous to the available `_irred_abelian(G::Oscar.GAPGroup)`
which computes the `QQ`-irreducibles of finite abelian groups.

(The functionality will be mentioned in a forthcoming paper.)

fieker commented on experimental/GModule/src/GModule.jl:
```diff
@@ -832,6 +833,74 @@ function _irred_abelian(G::Oscar.GAPGroup)
+    # (In the end, we will construct modules,
+    # there is no need to create an epimorphism to the largest p'-factor.)
+    function pprime_part(g, p)
+      v, _ = remove(order(g), p)
```
Do you know `ppio`? `ppio(a, b) = gcd(a, b^infty)`, so here `ppio(q, p) = p^k` for `k = valuation(q, p)`

ThomasBreuer replied:
Thanks.
`ppio` is a very suggestive name.
(Searching for it in the Oscar manual yields hundreds of matches, most (all?) of them have nothing to do with `ppio`.)

fieker commented on experimental/GModule/src/GModule.jl:
```diff
@@ -832,6 +833,74 @@ function _irred_abelian(G::Oscar.GAPGroup)
+    # The irred. `F`-repres. of `G` live already over `K`,
+    # we work over `K` and later write the matrices over `F`.
+    t = gcd(k, m)
+    K = GF(p^t)
```
I'd use `GF(p, m)` as the first thing `GF(p^k)` does is to "factor" to find p and k...

fieker commented on experimental/GModule/src/GModule.jl:
```diff
@@ -832,6 +833,74 @@ function _irred_abelian(G::Oscar.GAPGroup)
+    mp = embed(K, L)
+    z = Oscar.DiscLog.generator(L)
+    mat = companion_matrix(minpoly(z, mp))  # matrix over K, write over F
+    mmat = map_entries(embed(K, F), mat)^divexact(p^m-1, o)
```
I'm trying to understand the math, here it looks like `z` has order p^m-1, then the matrix has the same order, but is over K, the smallest field. Finally the matrix is coerced up and then powerd. Is this different from powering the generator (should be faster to power an element rather than a matrix?) or st least do the `nmap` powering in `K` rather than `F`?
For the current examples it probably does not matter at all.
If my guess is correct, then you don't need a generator, just an element of order `o` - which is much easier to get.

ThomasBreuer replied:
Yes, for an element from `dual(A)` of order `o`, we need an element of order `o` in an extension field of the given finite field. If we can get this without creating a generator of the multiplicative group of this field, this is fine; we only have to choose the *same* field element whenever we need one of order `o` (perhaps by caching the used elements in a dictionary), then the `dual(A)` magic will take care of the rest.

I guess the underlying math should be explained in the code.
The starting point for the new `_irred_abelian(G::Oscar.GAPGroup, F::FinField)` as well as for `_irred_abelian(G::Oscar.GAPGroup)` is that the image of an irreducible representation of an abelian group is cyclic, and that one gets the irreducible representations over a prescribed field as Galois sums of absolutely irreducible representations. Thus the representing matrices for each representation are powers of one fixed matrix of prescribed order `o`, and this matrix can be taken as the matrix of the `K`-linear map induced by an element of order `o` in the field `L`, where `L` is the smallest field that contains elements of order `o` and `K` is the intersection of `L` with the field `F` over which we want to realize the representations.

fieker replied:
Thinking about it some more: I am going to need an exp(G)-th root of unity (or possible the prime to p part) all others will be powers of that one. It might be better to do this once and then use powers? And or: use the `algebraic_closure(F)` and "fix" the code to have a consistent root of unity?

ThomasBreuer replied:
Yes, if `e` denotes the `p`-prime part of `exponent(G)` then some abs. irred. representation of the abelian group `G` will need a primitive `e`-th root of unity. For any abs. irred. representation of `G`, a power `r` of this root will be needed, and the field extension for the Galois action is `L/intersect(L. F)` where `L` is the smallest field containing `r`.

All computations can be done in a field that contains a primitive `e`-th root, and at least some computations cannot be done in a smaller field. The unique place in the code where one has to specify the field extensions is the computation of the minimal polynomial (or of the representation matrix) of `r` w.r.t. `L/intersect(L. F)`, which varies for different `r`.

fieker commented on experimental/GModule/src/GModule.jl:
```diff
@@ -832,6 +833,74 @@ function _irred_abelian(G::Oscar.GAPGroup)
+      e = r.elt::QQFieldElem
+      return (mmat^Int(divexact(o, denominator(e))))^Int(numerator(e))
+    end
+
```
As above, if my reasoning is correct, the function `conv` should operate on the elements and then produce the matrix? In my original code probably as well now that I think about it?
I guess I copied too much from the abs_irred case...
Why do I do `(a^b)^c`? and not `a^(b*c)`? Any ideas? or is it too early for me?

ThomasBreuer replied:
I think we can change the `conv` functions as you suggest.

fieker replied:
the test seem to error due to lacking `multiplicities`?

ThomasBreuer replied:
Yes, that is a utility function which I forgot to add to the testfile.
(Or perhaps it would be of general interest: Take a vector of objects that can be compared with `==`. and return a vector or perhaps a dictionary of `value => multiplicity` pairs.)

ThomasBreuer committed:
make the tests pass

(the missing `multiplicities` function is not needed, use `multiset` instead)

fieker commented:
On Tue, Sep 30, 2025 at 01:46:41AM -0700, Thomas Breuer wrote:
> @ThomasBreuer commented on this pull request.
> 
> 
> 
> > +# `G` over the field `F` of size `p^k`
> +function _irred_abelian(G::Oscar.GAPGroup, F::FinField)
> +  @assert is_abelian(G)
> +  q = order(F)
> +  p = characteristic(F)
> +  k, _ = remove(q, p)
> +
> +  if mod(order(G), p) != 0
> +    GG = G
> +  else
> +    # Replace `G` by its Hall p'-subgroup,
> +    # such that the generators correspond to those of `G`.
> +    # (In the end, we will construct modules,
> +    # there is no need to create an epimorphism to the largest p'-factor.)
> +    function pprime_part(g, p)
> +      v, _ = remove(order(g), p)
> 
> Thanks.
> `ppio` is a very suggestive name.
Dan Bernstein: PrimePowersInOut
actually he defines ppi, ppo, ppg and pple. We combined ppi and ppo as
in
https://cr.yp.to/lineartime/dcba-19980108.pdf
page 11
Heinz Lueneburg calls it POLr...
> (Searching for it in the Oscar manual yields hundreds of matches, most (all?) of them have nothing to do with `ppio`.)
But ?ppio would have been successful...
> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/oscar-system/Oscar.jl/pull/5394#discussion_r2390376674
> You are receiving this because your review was requested.
> 
> Message ID: ***@***.***>

fieker commented:
On Tue, Sep 30, 2025 at 05:24:27AM -0700, Thomas Breuer wrote:
> @ThomasBreuer commented on this pull request.
> 
> 
> 
> > +    # we work over `K` and later write the matrices over `F`.
> +    t = gcd(k, m)
> +    K = GF(p^t)
> +
> +    mp = embed(K, L)
> +    z = Oscar.DiscLog.generator(L)
> +    mat = companion_matrix(minpoly(z, mp))  # matrix over K, write over F
> +    mmat = map_entries(embed(K, F), mat)^divexact(p^m-1, o)
> +    f = mR(r)
> +    M = free_module(F, orblen)
> +
> +    function conv(r)
> +      e = r.elt::QQFieldElem
> +      return (mmat^Int(divexact(o, denominator(e))))^Int(numerator(e))
> +    end
> +
> 
> Yes, that is a utility function which I forgot to add to the testfile.
> (Or perhaps it would be of general interest: Take a vector of objects that can be compared with `==`. and return a vector or perhaps a dictionary of `value => multiplicity` pairs.)
> 
MSet ... you can compare the MSets
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/oscar-system/Oscar.jl/pull/5394#discussion_r2391214651
> You are receiving this because your review was requested.
> 
> Message ID: ***@***.***>

ThomasBreuer committed:
improve `_irred_abelian(G::Oscar.GAPGroup)`

- Create powers of the root of unity, not of its `representation_matrix`.
- Call `representation_matrix` for each of the powers.
- Compute `z^(a*b)` not `(z^a)^b`.

This is apparently faster.

ThomasBreuer committed:
address comments

- use `ppio`
- in the loop over the elements of the dual group,
  cache the information that depends only on the root of unity in question;
  this way, make explicit that the the roots of unity are compatible
- compute `m^(a*b)` not `(m^a)^b`
- write the matrices over `F` as late as possible

ThomasBreuer commented:
@fieker
Concerning the idea to work with powers of finite field elements as long as possible, in order to avoid powering of matrices:
The idea is to take an element `z` in the field `L`, and to compute the matrix of the `K`-linear map of the `K`-vector space `L` that is induced by multiplication with `z`.

Since the minimal polynomial of `z` over `K` is known to be irreducible *and* since there is a function that computes this minimal polynomial, we can compute the companion matrix of this polynomial. This trick does in general not work for powers of `z`.

As far as I see, there is no `representation_matrix` (yet) for this situation; note that `K` is in general not a prime field.

(Logically, the easiest approach would be to create `L` as an extension field of the intended field `F`, but then we may have to work in larger fields than necessary.)

fieker commented:
On Wed, Oct 01, 2025 at 02:46:19AM -0700, Thomas Breuer wrote:
> ThomasBreuer left a comment (oscar-system/Oscar.jl#5394)
> 
> @fieker
> Concerning the idea to work with powers of finite field elements as long as possible, in order to avoid powering of matrices:
> The idea is to take an element `z` in the field `L`, and to compute the matrix of the `K`-linear map of the `K`-vector space `L` that is induced by multiplication with `z`.
> 
> Since the minimal polynomial of `z` over `K` is known to be irreducible *and* since there is a function that computes this minimal polynomial, we can compute the companion matrix of this polynomial. This trick does in general not work for powers of `z`.
> 
> As far as I see, there is no `representation_matrix` (yet) for this situation; note that `K` is in general not a prime field.
> 
> (Logically, the easiest approach would be to create `L` as an extension field of the intended field `F`, but then we may have to work in larger fields than necessary.)
> 
I played with it this morning. We'd need the representation_matrix wrt
to some fixed basis, ie. the powers of z.
The easiest should be sub - but we don't have it. One of these days we
need to improve the fin. fields.
So for the time being, we can't do anything better.
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/oscar-system/Oscar.jl/pull/5394#issuecomment-3355547024
> You are receiving this because you were mentioned.
> 
> Message ID: ***@***.***>
